Multiple systemd/dracut fixes/changes

c0d3z3r0 opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://github.com/zfsonlinux/zfs/wiki/Buildbot-Options
-->

### Motivation and Context / Description
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

1. Remove hard dependency on bash

    zfs-import-* services have a hard dependency on bash while not everyone
    has bash installed. At this point /bin/sh is sufficient, so use that.

2. Fix systemd-import services

    On debian systemd complains about missing /bin/awk because it actually
    is located at /usr/bin/awk. It is not a good idea to hardcode binary
    paths because linux distros use different paths. According to systemd's
    man page it is absolutely safe to miss paths for binary located at
    standard locations (/bin, /sbin, /usr/bin, ...).

    Further, replace this more or less complicated awk command by egrep.

3. Move dracut specifics to dracut mount unit

    Dracut depends on BOOTFS environment variable to be set after pool
    import. This dracut specific systemd ExecStartPost command should not be
    called for any non-dracut systems, so let's move it to a static systemd
    unit that gets pulled in by the generated mount unit.


~~*At the moment, the new systemd service does not get installed / packaged. I did not get this to work... any hint from you would be great :-)*~~ **Update: Fixed.**

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->

Tested WITHOUT dracut! Maybe someome with a dracut (test) setup can test this?

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

kpande commented:
we might want some but not all of these changes, can you break them into separate PRs for easier review?

c0d3z3r0 commented:
Well, all three commits modify the same line in the two unit files and depend on each...
If you insist, I can split them tomorrow. Just give me a hint, which of my changes you might not want to merge.



> Am 18.03.2019 um 00:53 schrieb kpande <notifications@github.com>:
> 
> we might want some but not all of these changes, can you break them into separate PRs for easier review?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

rlaager commented on etc/systemd/system/zfs-import-cache.service.in:
```diff
@@ -13,7 +13,7 @@ ConditionPathExists=@sysconfdir@/zfs/zpool.cache
 RemainAfterExit=yes
 ExecStart=@sbindir@/zpool import -c @sysconfdir@/zfs/zpool.cache -aN
-ExecStartPost=/bin/sh -c "/bin/systemctl set-environment BOOTFS=$(@sbindir@/zpool list -H -o bootfs | /bin/awk '$1 != \"-\" {print; exit}')"
+ExecStartPost=/bin/sh -c "/bin/systemctl set-environment BOOTFS=$(@sbindir@/zpool list -H -o bootfs | egrep -v '^-$')"
```
As far as I can tell, `grep` is sufficient. I don't see any reason it needs to be `egrep`.

c0d3z3r0 replied:
Indeed, thanks!

behlendorf commented on contrib/dracut/systemd/zfs-dracut-bootfs.service.in:
```diff
@@ -0,0 +1,10 @@
+
+[Service]
+Type=oneshot
+ExecStart=/bin/sh -c "/bin/systemctl set-environment BOOTFS=$(@sbindir@/zpool list -H -o bootfs | grep -v '^-$')"
```
I believe you should add `-m1` to the grep so it exits on first match (like the previous awk).

c0d3z3r0 commented:
@behlendorf -m1 does not work here because of the negative match. (AFAIK. I can‘t test this right now.)

When there are multiple matches, how should we know which one is the right one? Hoping the first one is always correct does not really sound right to me..


> Am 18.03.2019 um 22:09 schrieb Brian Behlendorf <notifications@github.com>:
> 
> @behlendorf commented on this pull request.
> 
> In contrib/dracut/systemd/zfs-dracut-bootfs.service.in:
> 
> > @@ -0,0 +1,10 @@
> +[Unit]
> +Description=Set BOOTFS environment for dracut
> +Documentation=man:zfs(8)
> +DefaultDependencies=no
> +After=zfs-import.target
> +Requires=zfs-import.target
> +
> +[Service]
> +Type=oneshot
> +ExecStart=/bin/sh -c "/bin/systemctl set-environment BOOTFS=$(@sbindir@/zpool list -H -o bootfs | grep -v '^-$')"
> I believe you should add -m1 to the grep so it exits on first match (like the previous awk).
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

cwedgwood commented:
> @behlendorf -m1 does not work here because of the negative match. (AFAIK. I can‘t test this right now.) When there are multiple matches, how should we know which one is the right one? Hoping the first one is always correct does not really sound right to me..

`head -1` perhaps?  it doesn't solve the 'right one' issue but at least would be consistent with what we have/had so far

behlendorf commented:
> at least would be consistent with what we have/had so far.

On a properly configured system there should only be one match.  But when that's not the case picking the first one at least allows us to do something sensible.  Let's at least keep the behavior consistent until we can do something better.  I'd be OK with `head -1`.

c0d3z3r0 commented:
I added `head -1` now. Packaging still does not work; the tests are complaining about that, too. :/

c0d3z3r0 commented:
The packaging issue is fixed now, too

rlaager commented:
> I added `head -1` now. Packaging still does not work; the tests are complaining about that, too. :/

This is a bit pedantic, but technically speaking, the `head -1` should be on the commit that changes it to `grep`. I see it's there on the last commit, but I missed it the first time through when reviewing.

c0d3z3r0 commented:
> > I added `head -1` now. Packaging still does not work; the tests are complaining about that, too. :/
> 
> This is a bit pedantic, but technically speaking, the `head -1` should be on the commit that changes it to `grep`. I see it's there on the last commit, but I missed it the first time through when reviewing.

Full ACK. Fixed

gdevenyi commented on contrib/dracut/systemd/zfs-dracut-bootfs.service.in:
```diff
@@ -0,0 +1,10 @@
+
+[Service]
+Type=oneshot
+ExecStart=/bin/sh -c "/bin/systemctl set-environment BOOTFS=$(@sbindir@/zpool list -H -o bootfs | grep -v '^-$' | head -1)"
```
``head -1`` isn't part of the POSIX standard, ``head -n 1`` is

https://www.unix.com/man-page/posix/1P/HEAD/

c0d3z3r0 replied:
Well, I agree, but maybe we shouldn't mix too much in this PR. Making everything POSIX conformant may be worth another PR?

~~~
contrib/dracut/02zfsexpandknowledge/module-setup.sh.in:    numfields=`head -1 /proc/self/mountinfo | awk '{print NF}'`
config/kernel.m4:			             2>/dev/null | grep -v obj | tail -1)
config/deb.am:	arch=`$(RPM) -qp $${name}-kmod-$${version}.src.rpm --qf %{arch} | tail -1`; \
config/deb.am:	arch=`$(RPM) -qp $${name}-dkms-$${version}.src.rpm --qf %{arch} | tail -1`; \
config/deb.am:	arch=`$(RPM) -qp $${name}-$${version}.src.rpm --qf %{arch} | tail -1`; \
config/deb.am:	pkg10=`ls python*-pyzfs-$${version}* | tail -1`; \
config/tgz.am:	arch=`$(RPM) -qp $${name}-kmod-$${version}.src.rpm --qf %{arch} | tail -1`; \
config/tgz.am:	arch=`$(RPM) -qp $${name}-$${version}.src.rpm --qf %{arch} | tail -1`; \
debian/zfs-dkms.postinst:	addrlen=$(head -1 /proc/kallsyms| grep -o '^ *[^ ]*' |wc -c)
debian/zfs-dkms.config:	addrlen=$(head -1 /proc/kallsyms|awk '{print $1}'|wc -c)
tests/zfs-tests/tests/functional/cli_user/misc/arc_summary_001_pos.ksh:log_must eval "arc_summary | head -1 > /dev/null"
tests/zfs-tests/tests/functional/zvol/zvol_common.shlib:				tail -1 | awk '{print $3}')
tests/zfs-tests/tests/functional/history/history_003_pos.ksh:typeset orig_md5=$(zpool history $spool | head -2 | md5sum | \
tests/zfs-tests/tests/functional/history/history_003_pos.ksh:typeset final_md5=$(head -2 $TMPFILE | md5sum | awk '{print $1}')
tests/zfs-tests/tests/functional/bootfs/bootfs_006_pos.ksh:	VAL=$(zpool get bootfs $POOL | tail -1 | awk '{print $3}' )
tests/zfs-tests/tests/functional/bootfs/bootfs_006_pos.ksh:	VAL=$(zpool get bootfs $POOL | tail -1 | awk '{print $3}' )
tests/zfs-tests/tests/functional/bootfs/bootfs_003_pos.ksh:	RES=$(zpool get bootfs $POOL | tail -1 | awk '{print $3}' )
tests/zfs-tests/tests/functional/cli_root/zfs_create/zfs_create_012_pos.ksh:ZFS_VERSION=$(zfs upgrade | head -1 | awk '{print $NF}' \
tests/zfs-tests/tests/functional/cli_root/zpool_import/zpool_import.kshlib:		    awk '$3=="C" {print $1}' | tail -1)
tests/zfs-tests/tests/functional/cli_root/zfs_clone/zfs_clone_010_pos.ksh:		actual_clone=$(zfs list -t all -o clones $snapshot | tail -1)
tests/zfs-tests/tests/functional/cli_root/zfs_clone/zfs_clone_010_pos.ksh:char_count=$(echo "$clone_list" | tail -1 | wc | awk '{print $3}')
tests/zfs-tests/tests/functional/cli_root/zfs_clone/zfs_clone_007_pos.ksh:ZFS_VERSION=$(zfs upgrade | head -1 | awk '{print $NF}' \
tests/zfs-tests/tests/functional/cli_root/zfs_send/zfs_send_006_pos.ksh:		typeset total_size=$(zfs send $option $snapshot 2>&1 | tail -1)
tests/zfs-tests/tests/functional/cli_root/zfs_send/zfs_send_006_pos.ksh:		     2>&1 | tail -1)
tests/zfs-tests/tests/functional/cli_root/zpool_upgrade/zpool_upgrade_001_pos.ksh:log_must eval "zpool upgrade -v | head -1 | grep 'feature flags'"
tests/zfs-tests/tests/functional/slog/slog_015_neg.ksh:ORIG_TIMEOUT=$(get_tunable zfs_commit_timeout_pct | tail -1 | awk '{print $NF}')
tests/zfs-tests/include/libtest.shlib:			unit cyl print | head -3 | tail -1 | \
tests/zfs-tests/include/libtest.shlib:		prop_val=$(zpool get -pH $prop $pool 2>/dev/null | tail -1 | \
cmd/vdev_id/vdev_id:	PHY=`ls -d $port_dir/phy* 2>/dev/null | head -1 | awk -F: '{print $NF}'`
cmd/zed/zed.d/zed-functions.sh:        | tail -1 | cut -d\; -f1)"
scripts/zloop.sh:        printf "%s" "$(ls -tr1 $coreglob 2> /dev/null | head -1)"
~~~

gdevenyi replied:
Sure, makes sense, I just happened by this while reading latest things in the repo.

behlendorf replied:
I agree, we don't want to make any other unrelated changes in this PR.  But let's use the standard POSIX version in `zfs-dracut-bootfs.service.in` while we're already making this change.

Aside from that this PR looks functionally correct to me, but we're going to want to test it with a dracut setup.

rlaager commented:
In thinking about this some more, a `grep | head` is ripe for replacement by `awk`. This avoids a second `fork()`. Why not just keep the existing approach, but remove the hard-coded `/bin/`?

c0d3z3r0 commented:
@rlaager
> In thinking about this some more, a `grep | head` is ripe for replacement by `awk`. This avoids a second `fork()`. Why not just keep the existing approach, but remove the hard-coded `/bin/`?

`grep | head` is still better readable than `awk '$1 != \"-\" {print; exit}')` and two forks instead of one is not *that* big difference.

@rlaager @behlendorf 
I just read the `grep` man page again and realized I was wrong.... `-m1` **does** work with a negative match. I simply stumbled over that double-negation in `When the -v or --invert-match option is also used, grep stops after outputting NUM **non-matching** lines.`, where "non-matching" means lines that do ***not*** match the (inverted match!) regex.
I will change this to `grep -v -m 1 ...`

kpande commented on contrib/dracut/Makefile.am:
```diff
@@ -1,3 +1,4 @@
-SUBDIRS = 02zfsexpandknowledge 90zfs
+SUBDIRS = 02zfsexpandknowledge 90zfs $(ZFS_INIT_SYSTEMD)
+DIST_SUBDIRS = 02zfsexpandknowledge 90zfs systemd
```
why did you separate it into a non-numbered directory?

c0d3z3r0 replied:
Since this systemd service file is dracut specific and I would expect it to be installed with the zfs-dracut package, it seemed logic to me to move it to the dracut directory. Maybe this should be even moved to `90zfs`?

c0d3z3r0 replied:
Since this systemd service file is dracut specific and I would expect it to be installed with the zfs-dracut package, it seemed logic to me to move it to the dracut directory. Maybe this should be moved to `90zfs`?

kpande replied:
why not `etc/system/systemd`?

rlaager replied:
`/etc/systemd/system` is for the system administrator. Packages should install to `/usr/lib/systemd/system` or `/lib/systemd/system`, depending on the distro. The correct path can be obtained via `pkg-config`.

c0d3z3r0 replied:
@rlaager kpande meant `etc/systemd/system` in the repo which gets installed to `@systemdunitpath@`

@kpande As I said, this is dracut specific. `etc/systemd/system` are general zfs units. Why is the generator in `contrib/dracut` when it is systemd?

kpande replied:
@rlaager then maybe [this](https://github.com/zfsonlinux/zfs/tree/master/etc/systemd/system) should not be here?

kpande replied:
@c0d3z3r0 I've actually got no idea why there is a systemd generator in `contrib/dracut`.

c0d3z3r0 replied:
@kpande Great :D   I'd be ok with moving the unit file to the systemd tree, while keeping it installed only with `zfs-dracut`. I just want to see if @behlendorf has some thoughts on this.

kpande replied:
maybe @prometheanfire can chime in since i think he is the one who created that generator

prometheanfire replied:
I could have easily created it wrong.  I don't create them every day.  It certainly sounds wrong if it's trying to run systemctl things on non-systemd systems though.

kpande replied:
ok, if we can i'd like to keep systemd generators in one place, as it may not end up being dracut specific. maybe @c0d3z3r0 can move them both into `etc/systemd` or as @rlaager suggested maybe that also needs its own issue to then move everything into the path provided by the system.

c0d3z3r0 replied:
> ok, if we can i'd like to keep systemd generators in one place, as it may not end up being dracut specific. maybe @c0d3z3r0 can move them both into `etc/systemd` 

Ok

> or as @rlaager suggested maybe that also needs its own issue to then move everything into the path provided by the system.

No, that is not needed since it is only the tree structure in git which is misleading. The files do end up in `/usr/lib/....` or `/lib/....` based on `configure`'s arguments

c0d3z3r0 replied:
Hm. That generator is only called from dracut. It is directly put to the initramfs without ever being called on a fully booted system.... https://github.com/zfsonlinux/zfs/blob/master/contrib/dracut/90zfs/module-setup.sh.in#L67

I'm not sure if it is a good idea to put a generator to  `/lib`/`usr/lib` that is never called, except from within dracut.

kpande replied:
ok, I would like for us to find some path for dracut generators other than the module directory. it is not immediately necessary so long as all of this works on dracut+openrc|systemd.

c0d3z3r0 replied:
I looked a bit through other dracut modules and found this one for example: [nbd](https://git.kernel.org/pub/scm/boot/dracut/dracut.git/tree/modules.d/95nbd)
The generator is in the module directory, so 90zfs should be the right place for both the generator and the unit file

c0d3z3r0 replied:
Here is another one https://git.kernel.org/pub/scm/boot/dracut/dracut.git/tree/modules.d/98dracut-systemd/emergency.service

c0d3z3r0 replied:
I just pushed my changes. IMHO that is the right way :-)

prometheanfire commented on contrib/dracut/systemd/zfs-dracut-bootfs.service.in:
```diff
@@ -0,0 +1,10 @@
+
+[Service]
+Type=oneshot
+ExecStart=/bin/sh -c "/bin/systemctl set-environment BOOTFS=$(@sbindir@/zpool list -H -o bootfs | grep -m1 -v '^-$')"
```
I imagine the awk as used before may be faster / better not too sure.  It was a while ago, but I do remember trying things like grep.  It's possible I just didn't find the right combo.

c0d3z3r0 replied:
~~~
c0d3@z3r0:[~]# time echo -e '-\n-\nrpool/ROOT\ndpool/ROOT' | grep -m1 -v '^-$'
rpool/ROOT

real	0m0,003s
user	0m0,002s
sys	0m0,001s
c0d3@z3r0:[~]# time echo -e '-\n-\nrpool/ROOT\ndpool/ROOT' | awk '$1 != "-" {print; exit}'
rpool/ROOT

real	0m0,004s
user	0m0,003s
sys	0m0,001s
~~~

There should be no noticable difference in boot speed. That is absolutely negligible :-)

behlendorf commented on configure.ac:
```diff
@@ -127,6 +127,7 @@ AC_CONFIG_FILES([
 	contrib/dracut/Makefile
 	contrib/dracut/02zfsexpandknowledge/Makefile
 	contrib/dracut/90zfs/Makefile
+	contrib/dracut/systemd/Makefile
```
You can drop this now.

behlendorf commented on contrib/dracut/90zfs/Makefile.am:
```diff
@@ -9,6 +9,9 @@ pkgdracut_SCRIPTS = \
 	zfs-lib.sh
 
+pkgdracut_DATA = \
+	zfs-dracut-bootfs.service.in
```
Putting this in `DATA` makes sense, but I think you need to remove the `.in` suffix since we want to install the generated version.  Similar to `SCRIPTS` above.

c0d3z3r0 replied:
Oops... copy'n'paste mistake..

behlendorf commented on contrib/dracut/90zfs/module-setup.sh.in:
```diff
@@ -98,6 +98,7 @@ install() {
 				type mark_hostonly >/dev/null 2>&1 && mark_hostonly @systemdunitdir@/zfs-import-$_item.service
 			fi
 		done
+		inst "${moddir}/zfs-env-bootfs.service" "${systemdsystemunitdir}/zfs-env-bootfs.service"
```
Should this be `inst_simple`?

c0d3z3r0 replied:
I took this from dracut's [man page](http://man7.org/linux/man-pages/man7/dracut.modules.7.html) which only names `inst <src> [<dst>]`. [dracut-init.sh](https://git.kernel.org/pub/scm/boot/dracut/dracut.git/tree/dracut-init.sh#n182) defines both but there is not much difference, at least at this point... both should be OK here

Rudd-O commented on contrib/dracut/90zfs/zfs-env-bootfs.service.in:
```diff
@@ -0,0 +1,10 @@
+Documentation=man:zfs(8)
+DefaultDependencies=no
+After=zfs-import.target
+Requires=zfs-import.target
```
This is not correct and will break systems.  This should be ordered `After=zfs-import-*.service` and `Before=zfs-import.target`.

Rudd-O replied:
It is imperative that the `BOOTFS` variable be set *before* reaching `zfs-import.target` because other parts of the initial RAM disk (and quite likely third-party programs too) have come to depend on that assumption.

c0d3z3r0 replied:
No. You need to import a pool first before reading the bootfs parameter. This was initially a ExecStartPost, so After is correct.

c0d3z3r0 replied:
Ah, wait... yes, you are right.. everything should be inside the import target

c0d3z3r0 replied:
@Rudd-O I looked at this again. `zfs-env-bootfs` is pulled in by the generated `zfs-enhancement.conf` [here](https://github.com/zfsonlinux/zfs/pull/8510/files/c1fb1e7b1493e935078c3ea0fd246e8afdfa25db#diff-6809a8a118df0d43b694eb6227b25cc7R47)

`zfs-enhancement.conf` loads after `zfs-import.target` and that is the only part of the initramfs that depends on `BOOTFS`.

While I believe that this would *not* break anything, I move `zfs-env-bootfs.service` to `zfs-import.target`, nevertheless, because this way we make sure we do not change the previous behaviour.

Rudd-O commented on etc/systemd/system/zfs-import-cache.service.in:
```diff
@@ -13,7 +13,6 @@ ConditionPathExists=@sysconfdir@/zfs/zpool.cache
 Type=oneshot
 RemainAfterExit=yes
 ExecStart=@sbindir@/zpool import -c @sysconfdir@/zfs/zpool.cache -aN
```
A nice-to-have here would be to import *only* the pool specified in `root=` (`/proc/cmdline`) if the user has decided to do that.  This prevents importing all available and importable pools unnecessarily.  I am going to assume this suggestion does not belong in this changeset.

c0d3z3r0 replied:
Isn‘t that possible already? The generator decides between root=zfs:auto and others

c0d3z3r0 commented:
No. You need to import a pool first before reading the bootfs parameter

> Am 26.03.2019 um 13:00 schrieb Rudd-O <notifications@github.com>:
> 
> @Rudd-O commented on this pull request.
> 
> In etc/systemd/system/zfs-import-cache.service.in:
> 
> > @@ -13,7 +13,6 @@ ConditionPathExists=@sysconfdir@/zfs/zpool.cache
>  Type=oneshot
>  RemainAfterExit=yes
>  ExecStart=@sbindir@/zpool import -c @sysconfdir@/zfs/zpool.cache -aN
> A nice-to-have here would be to import only the pool specified in root= (/proc/cmdline) if the user has decided to do that. This prevents importing all available and importable pools unnecessarily. I am going to assume this suggestion does not belong in this changeset.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

c0d3z3r0 commented:
Isn‘t that possible already? The generator decides between root=zfs:auto and others

> Am 26.03.2019 um 13:00 schrieb Rudd-O <notifications@github.com>:
> 
> @Rudd-O commented on this pull request.
> 
> In etc/systemd/system/zfs-import-cache.service.in:
> 
> > @@ -13,7 +13,6 @@ ConditionPathExists=@sysconfdir@/zfs/zpool.cache
>  Type=oneshot
>  RemainAfterExit=yes
>  ExecStart=@sbindir@/zpool import -c @sysconfdir@/zfs/zpool.cache -aN
> A nice-to-have here would be to import only the pool specified in root= (/proc/cmdline) if the user has decided to do that. This prevents importing all available and importable pools unnecessarily. I am going to assume this suggestion does not belong in this changeset.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

behlendorf commented:
I've cherry picked the first two simple commits in to master since there's no reason to hold them up any longer.  @c0d3z3r0 would you mind rebasing only your last commit on master while we allow some time for additional reviewers.  Thanks.

e03b25a56443 Fix systemd-import services
3b2618927c66 Remove hard dependency on bash

rlaager commented:
I removed my stale review. I don't see anything that jumps out at me as wrong, but I am not familiar with dracut, so I can't give a useful approval here.

c0d3z3r0 commented:
> tested fine

Thank you!

> 
> For some reason I didn't have zfs.target enabled and that caused zfs-mount to not start. I don't see a change in this commit that would have done that though (probably one of the other dracut commits)

Do you mean zfs.target in the rootfs / after initramfs (`etc/systemd/system/zfs.target.in`)? I could not find any usage of `zfs.target` in `contrib/dracut`

prometheanfire commented:
ya, after initram
